### PR TITLE
JFG-1111: add global context to scenario

### PR DESCRIPTION
**Pull request by anton-antonenko**

### CODE REVIEW

File: src/main/java/com/griddynamics/scenario/ExampleUserScenarioProvider.java [hunk 1]
```diff
@@ -46,6 +47,9 @@ public ExampleUserScenarioProvider() {
                 .addStep(JHttpUserScenarioStep.builder(STEP_3_ID)
                         .withDisplayName("Step #3")
                         .withQuery(new JHttpQuery().get().path("/response-headers?key=val"))
+                        // global context can be changed here
+                        .withPreProcessGlobalContextFunction((prevStep, context) ->
+                                context.withGlobalHeaders(prevStep.getResponse().getHeaders()))
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/scenario/ExampleUserScenarioProvider.java]: more comment + curly braces
```
// global context can be changed here. E.g. we can add headers to all requests in this scenario
.withPreProcessGlobalContextFunction((prevStep, context) -> {
                  context.withGlobalHeaders(prevStep.getResponse().getHeaders())
})
```

File: src/main/java/com/griddynamics/scenario/jagger/JHttpUserScenarioInvoker.java [hunk 2]
```diff
@@ -26,29 +25,41 @@ public JHttpUserScenarioInvocationResult invoke(Void nothing, JHttpUserScenario
...
 
-            // use global endpoint for step if present
-            if (globalEndpoint != null)
-                userScenarioStep.setEndpoint(globalEndpoint);
-            // check endpoint for null
-            if (globalEndpoint == null && userScenarioStep.getEndpoint() == null)
-                throw new IllegalArgumentException("Endpoint must not be null! Please, set global endpoint or set endpoint for every step.");
-
             // Pre process step: user actions executed before request
+            userScenarioStep.preProcessGlobalContext(previousStep, localScenarioContext);
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/scenario/jagger/JHttpUserScenarioInvoker.java:43]: is localScenarioContext immutable in this case?

Comment (anton-antonenko) [reply src/main/java/com/griddynamics/scenario/jagger/JHttpUserScenarioInvoker.java:43]: no. should it? this step is needed for updating context using previous step, as I understood.

Comment (dlatnikov) [reply src/main/java/com/griddynamics/scenario/jagger/JHttpUserScenarioInvoker.java:43]: it is correct. please see my other comment for this class

File: src/main/java/com/griddynamics/scenario/jagger/JHttpUserScenarioInvoker.java [hunk 3]
```diff
@@ -26,29 +25,41 @@ public JHttpUserScenarioInvocationResult invoke(Void nothing, JHttpUserScenario
         Boolean scenarioSucceeded = true;
         JHttpUserScenarioStep previousStep = null;
         List<JHttpUserScenarioStepInvocationResult> stepInvocationResults = new ArrayList<>();
-        JHttpEndpoint globalEndpoint = scenario.getGlobalEndpoint();
 
-        for (JHttpUserScenarioStep userScenarioStep : scenario.userScenarioSteps) {
+        JHttpScenarioGlobalContext localScenarioContext = scenario.getScenarioGlobalContext();
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/scenario/jagger/JHttpUserScenarioInvoker.java]: we need to make a copy here
scenario.getScenarioGlobalContext() should be not changed during scenario execution
we should start scenario with the same values every time

File: src/main/java/com/griddynamics/scenario/jagger/CopyUtil.java [hunk 4]
```diff
@@ -43,4 +44,12 @@ public static JHttpResponse copyOf(JHttpResponse jHttpResponse) {
         return new JHttpResponse(jHttpResponse.getStatus(), jHttpResponse.getBody(), jHttpResponse.getHeaders());
     }
 
+    public static HttpHeaders copyOf(HttpHeaders headers) {
+        if (headers == null)
+            return null;
+        HttpHeaders headersCopy = new HttpHeaders();
+        headersCopy.putAll(headers);
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/scenario/jagger/CopyUtil.java]: is it really making a copy?

**Comment by a-badaev** [comment]
'Default' label  fits better to me that 'Global' what do you think guys?

File: src/main/java/com/griddynamics/scenario/jagger/CopyUtil.java (hunk 4 above)
Comment (anton-antonenko) [reply src/main/java/com/griddynamics/scenario/jagger/CopyUtil.java]: now yes - check out last commit

**Review by a-badaev — APPROVED** [review]

**Review by dlatnikov — APPROVED** [review]

**Pull request merged**